rules/cache/cache.go

```diff
@@ -0,0 +1,378 @@
+	// is now in effect, so all the old cached results are now invalid, therefore it's
+	// preferrable to invalidate everything to save the overhead of validating old results.
+	if invalidate {
+		results = c.invalidateWithLock(namespace, idHash, results)
+	}
+	res := results.source.Match(id)
+	newElem := &element{namespace: namespace, idHash: idHash, result: res}
+	newElem.SetExpiry(c.newExpiry(c.nowFn()))
+	results.elems[idHash] = newElem
+	// NB(xichen): we don't evict until the number of cached items goes
+	// above the capacity by at least the eviction batch size to amortize
+	// the eviction overhead.
+	if newSize := c.add(newElem); newSize > c.capacity+c.evictionBatchSize {
+		c.notifyEviction()
+	}
+	return res
+}
+
+func (c *cache) add(elem *element) int {
+	c.list.Lock()
+	c.list.PushFront(elem)
+	size := c.list.Len()
+	c.list.Unlock()
+	return size
+}
+
+func (c *cache) promote(now time.Time, elem *element) {
+	c.list.Lock()
+	// Bail if someone else got ahead of us and promoted this element.
+	if !elem.ShouldExpire(now) {
+		c.list.Unlock()
+		return
+	}
+	// Otherwise proceed with promotion.
+	elem.SetExpiry(c.newExpiry(now))
+	c.list.MoveToFront(elem)
+	c.list.Unlock()
+}
+
+func (c *cache) invalidateWithLock(namespace string, idHash mid.Hash, results results) results {
+	if c.invalidationMode == InvalidateAll {
+		c.toDelete = append(c.toDelete, results.elems)
+		c.notifyDeletion()
+		results.elems = make(elemMap)
+		c.namespaces[namespace] = results
+	} else {
+		elem := results.elems[idHash]
+		delete(results.elems, idHash)
+		c.list.Lock()
+		c.list.Remove(elem)
+		c.list.Unlock()
+	}
+	return results
+}
+
+func (c *cache) evict() {
+	defer c.wgWorker.Done()
+
+	for {
+		select {
+		case <-c.evictCh:
+			c.doEvict()
+		case <-c.closedCh:
+			return
+		}
+	}
+}
+
+func (c *cache) doEvict() {
+	c.Lock()
+	c.list.Lock()
+	for c.list.Len() > c.capacity {
+		elem := c.list.Back()
+		c.list.Remove(elem)
+		// NB(xichen): the namespace owning this element may have been deleted,
+		// in which case we simply continue. This is okay because the deleted element
+		// will be marked as deleted so when the deletion goroutine sees and tries to
+		// delete it again, it will be a no op, at which point it will be removed from
+		// the owning map as well.
+		results, exists := c.namespaces[elem.namespace]
+		if !exists {
+			continue
+		}
+		delete(results.elems, elem.idHash)
+	}
+	c.list.Unlock()
+	c.Unlock()
+}
+
+func (c *cache) delete() {
+	defer c.wgWorker.Done()
+
+	for {
+		select {
+		case <-c.deleteCh:
+			c.doDelete()
+		case <-c.closedCh:
+			return
+		}
+	}
+}
+
+func (c *cache) doDelete() {
+	c.Lock()
+	if len(c.toDelete) == 0 {
+		c.Unlock()
+		return
+	}
+
+	// NB(xichen): add pooling if deletion happens frequent enough.
+	toDelete := c.toDelete
+	c.toDelete = nil
+	c.Unlock()
+
+	deleted := 0
+	c.list.Lock()
+	for _, elems := range toDelete {
+		for id, elem := range elems {
+			delete(elems, id)
```

Comment:
Is it worth deleting `id` from `elems` considering its unreachable after `doDelete(...)`?